README.md
此案例参考flappy_search_bar_fork 和 FultterIntroduction 控件改写而成   （显示联系人列表样式尚未完成）部分尚未改写完毕，更新中

flappy_search_bar_fork
==============================
This is fork of flappy_search_bar
----------------------------------------
A SearchBar widget handling most of search cases.

## Usage

To use this plugin, add flappy_search_bar as a dependency in your pubspec.yaml file.

### Example

```
@override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: SearchBar<Post, String>(
            searchBarPadding: EdgeInsets.symmetric(horizontal: 10),
            headerPadding: EdgeInsets.symmetric(horizontal: 10),
            listPadding: EdgeInsets.symmetric(horizontal: 10),
            getListData: _getData,
            searchType: SearchType.search,
            onChange: _getALlPosts,
            searchBarController: _searchBarController,
            textAligin: textAligin,
            textPaddingLeft: 30,
            // focusNode: focusNode,
            focusCall: () {
              setState(() {
                print('focuscall');
                textAligin = TextAlign.left;
                cancellationWidgetWidth = 120;
              });
            },
            hintText: '搜索',
            hintStyle: TextStyle(
              textBaseline: TextBaseline.alphabetic,
              fontSize: 16,
            ),
            placeHolder: Text("placeholder"),
            cancellationWidget: Text("Cancel"),
            cancellationWidgetWidth: cancellationWidgetWidth,
            emptyWidget: Text("empty"),
            改写后到属性 (hoverListProperty,scaledTileProperty只选其一)
            hoverListProperty: HoverListProperty(itemCounts: [5],xx,xx,),
            
            scaledTileProperty: ScaledTitleProperty(
              indexedScaledTileBuilder: (int index) => ScaledTile.extent(1, 60),
              onItemFound: (post, int index) {
                return Container(
                  color: Colors.lightBlue,
                  child: ListTile(
                    title: Text((post.title)),
                    // isThreeLine: true,
                    subtitle: Text(post.body),
                    onTap: () {
                      Navigator.of(context).push(
                          MaterialPageRoute(builder: (context) => Detail()));
                    },
                  ),
                );
              },
            ),
            //这两个属性属性被改写到一个属性里边
            // indexedScaledTileBuilder: (int index) =>ScaledTile.extent(1, 60),
            //indexedScaledTileBuilder: (int index) => ScaledTile.count(1, index.isEven ? 2 : 1),
            
            // onItemFound: (post, int index) {
            //   return Container(
            //     color: Colors.lightBlue,
            //     child: ListTile(
            //       title: Text((post.title)),
            //       // isThreeLine: true,
            //       subtitle: Text(post.body),
            //       onTap: () {
            //         Navigator.of(context).push(
            //             MaterialPageRoute(builder: (context) => Detail()));
            //       },
            //     ),
            //   );
            // },
            
            
            
            
            header: Row(
              children: <Widget>[
                RaisedButton(
                  child: Text("Sort"),
                  onPressed: () {
                    _searchBarController.sortList((Post a, Post b) {
                      return a.body.compareTo(b.body);
                    });
                  },
                ),
                RaisedButton(
                  child: Text("Desort"),
                  onPressed: () {
                    _searchBarController.removeSort();
                  },
                ),
                RaisedButton(
                  child: Text("Replay"),
                  onPressed: () {
                    isReplay = !isReplay;
                    _searchBarController.replayLastSearch();
                  },
                ),
                RaisedButton(
                    child: Text("Random"),
                    onPressed: () {
                      // cancellationWidgetWidth = 0; //random.nextInt(250) + 50.0;
                      // textPaddingLeft = 250;
                      // focusNode = FocusNode();
                      setState(() {});
                    })
              ],
            ),
            onCancelled: (String value) {
              print("Cancelled triggered");

              setState(() {
                textAligin = TextAlign.center;
                cancellationWidgetWidth = 0; //random.nextInt(250) + 50.0;
              });
            },
            mainAxisSpacing: 0,
            crossAxisSpacing: 10,
            crossAxisCount: 1,
            onItemLenovoFound: (item, int index) {
              return Text('item ---- ${index}');
            }, //2,
            
          ),
      ),
    );
  }
```

### Try it

A sample app is available to let you try all the features ! :)

### Warning

If you want to use a SearchBarController in order to do some sorts or filters, PLEASE put your instance of SearchBarController in a StateFullWidget.

If not, it will not work properly.

If you don't use an instance of SearchBarController, you can keep everything in a StateLessWidget !

### Parameters

| Name  | Type | Usage | Required | Default Value |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| onSearch   | Future<List<T>> Function(String text) | Callback giving you the text to look for and asking for a Future  | yes  | - |
| onItemFound| Widget Function(T item, int index) | Callback letting you build the widget corresponding to each item| yes| - |
| suggestions  |  List<T> | Potential fist list of suggestions (when no request have been made)  | no| [] |
| searchBarController  |  SearchBarController | Enable you to sort and filter your list  | no | default controller |
| searchBarStyle  |  SearchBarStyle | Syle to customize SearchBar  | no | default values on bottom tab |
| buildSuggestions| Widget Function(T item, int index) | Callback called to let you build Suggestion item (if not provided, the suggestion will have the same layout as the basic item)  | no| null|
| minimumChars  |  int | Minimum number of chars to start querying  | no| 3 |
| onError  |  Function(Error error) | Callback called when an error occur runnning Future | no| null |
| debounceDuration  | Duration | Debounce's duration | no| Duration(milliseconds: 500) |
| loader  | Widget | Widget that appears when Future is running | no| CircularProgressIndicator() |
| emptyWidget  | Widget | Widget that appears when Future is returning an empty list | no| SizedBox.shrink() |
| icon  | Widget | Widget that appears on left of the SearchBar | no| Icon(Icons.search) |
| hintText  | String | Hint Text | no| "" |
| hintStyle  | TextStyle | Hint Text style| no| TextStyle(color: Color.fromRGBO(142, 142, 147, 1)) |
| iconActiveColor  | Color | Color of icon when active | no| Colors.black |
| textStyle  | TextStyle | TextStyle of searched text | no| TextStyle(color: Colors.black) |
| cancellationWidget  | Widget | Widget shown on right of the SearchBar | no| Text("Cancel") |
| onCancelled  | VoidCallback | Callback triggered on cancellation's button click | no| null |
| crossAxisCount  | int | Number of tiles on cross axis (Grid) | no| 2 |
| shrinkWrap  | bool | Wether list should be shrinked or not (take minimum space) | no| true |
| scrollDirection  | Axis | Set the scroll direction | no| Axis.vertical |
| mainAxisSpacing  | int | Set the spacing between each tiles on main axis | no| 10 |
| crossAxisSpacing  | int | Set the spacing between each tiles on cross axis | no| 10 |
| indexedScaledTileBuilder  | IndexedScaledTileBuilder | Builder letting you decide how much space each tile should take | no| (int index) => ScaledTile.count(1, index.isEven ? 2 : 1) |  
| searchBarPadding  | EdgeInsetsGeometry | Set a padding on the search bar | no| EdgeInsets.symmetric(horizontal: 10) |
| headerPadding  | EdgeInsetsGeometry | Set a padding on the header | no| EdgeInsets.symmetric(horizontal: 10) |
| listPadding  | EdgeInsetsGeometry | Set a padding on the list | no| EdgeInsets.symmetric(horizontal: 10) |
  
### SearchBar default SearchBarStyle

| Name  | Type | default Value |
| ------------- | ------------- | ------------- |
| backgroundColor  | Color  | Color.fromRGBO(142, 142, 147, .15)  |
| padding  | EdgeInsetsGeometry  | EdgeInsets.all(5.0)  |
| borderRadius  | BorderRadius  | BorderRadius.all(Radius.circular(5.0))})  |





CHANGELOG.md
## [1.0.0] - 25/07/2019

* Initial Release

## [1.1.0] - 25/07/2019

* Adding Header, Controller and Styling

This version enable you to ass a Header (Widget that will be placed between SearchBar and ListView.
It also gives you the opportunity to have a Controller which is enable to sort or filter the list at any time.
Finally, giving a SearchBarStyle, you will be able to change the SearchBar's styling.

## [1.1.1] - 25/07/2019

* Fixing bug on hot reload and on Controller state


## [1.2.0] - 26/07/2019

* Future are now cancelled if another request is made and prev was not completed yet
* Increased a little bit the "cancellation" widget click's surface

## [1.3.0] - 29/07/2019

* SearchBarController is now enable to replay last search

## [1.3.1] - 29/07/2019

* Fix API replay's method's name

## [1.4.0] - 04/10/2019

* Add possibility to display items in a Grid
* Possibility to customize the space taken by each individual tile 

## [1.4.1] - 04/10/2019

* Add possibility to customize padding on List, Header and Search bar

## [1.5.1] - 14/01/2020

* Add callback on cancellation's button click

## [1.5.2] - 14/01/2020

* Fix README by adding cancellation's callback documentation

## [1.6.2] - 15/01/2020

* Add possibility to clear everything with the controller

## [1.7.2] - 24/01/2020

* Change cancellation widget in order to take a `Widget` instead of just a `Text` to make it more customizable

## [1.7.3] - 21/06/2020

* Fork flappy_search_bar: flappy_search_bar_search
* Add cancellation widget width to make it more customizable
* Add "Random" button in example app that change cancellation widget width randomly

## [1.7.4] - 21/06/2020

* Allow customize SearchBar Border
* Allow customize SearchBar Height

lib/index_bar.dart
import 'package:flutter/material.dart';
import 'const.dart';

class IndexBar<T> extends StatefulWidget {
  final int Function(T item1, T item2) sort;
  final String Function(T item) byValue;
  final List<T> dataList;
  final List<String> indexWordListOrBeforeList;
  //创建索引条回调
  final void Function(String str) indexBarCallBack;
  final Color indexTextColor;
  final ImageProvider<Object> assetImg;

  IndexBar(
      {this.sort,
      this.byValue,
      this.dataList,
      this.indexWordListOrBeforeList,
      this.indexBarCallBack,
      this.indexTextColor,
      this.assetImg});

  @override
  _IndexBarState createState() => _IndexBarState();
}

int getIdex(BuildContext context, Offset globalPosition, List index_word) {
//  拿到box
  RenderBox box = context.findRenderObject();
//  拿到y值
  double y = box.globalToLocal(globalPosition).dy;
//  算出字符高度  box 的总高度 / 2 / 字符开头数组个数
  var itemHeight = ScreenHeight(context) / 2 / index_word.length;
  //算出第几个item，并且给一个取值范围   ~/ y除以item的高度取整  clamp 取值返回 0 -
  int index = (y ~/ itemHeight).clamp(0, index_word.length - 1);

  print('现在选中的是${index_word[index]}');
  return index;
}

final List<String> _index_word = [];

class _IndexBarState<T> extends State<IndexBar<T>> {
  Color _bkColor = Color.fromRGBO(1, 1, 1, 0.0);
  Color _textColor = Colors.black;
  double _indicatorY = 0.0; //悬浮窗位置
  String _indicatorText = 'A'; //显示的字母
  bool _indocatorHidden = true; //是否隐藏悬浮窗
  ImageProvider<Object> assetImg;

//  排序后的数组
  final List<T> _listDatas = [];
  @override
  void initState() {
    // TODO: implement initState
    super.initState();
//----------------------- 1 -------------------------------
//    1、根据实际数据显示右侧bar
    // _index_word.add('🔍');
    // _index_word.add('☆');
    assetImg = widget.assetImg ?? AssetImage('images/气泡.png');
    if (null != widget.indexTextColor) {
      _textColor = widget.indexTextColor;
    }

    if (null != widget.indexWordListOrBeforeList) {
      _index_word.addAll(widget.indexWordListOrBeforeList);
    }
    if (widget.dataList.isNotEmpty && widget.dataList.length > 0) {
      _listDatas.addAll(widget.dataList);
      if (null != widget.sort) {
        //排序!
        _listDatas.sort((T a, T b) {
          return widget.sort(a, b);
        });
      }

//经过循环，将每一个头的首字母放入index_word数组
      if (null != widget.byValue) {
        for (int i = 0; i < _listDatas.length; i++) {
          if (i < 1 ||
              widget.byValue(_listDatas[i]) !=
                  widget.byValue(_listDatas[i - 1])) {
            _index_word.add(widget.byValue(_listDatas[i]));
          }
        }
      }
    }

    //----------------------- 2 -------------------------------
//    2、右侧bar显示全部字母
//    _index_word.addAll(INDEX_WORDS);
  }

  @override
  Widget build(BuildContext context) {
    List<Widget> words = [];
    for (int i = 0; i < _index_word.length; i++) {
      words.add(Expanded(
          child: Text(
        _index_word[i],
        style: TextStyle(fontSize: 10, color: _textColor),
      )));
    }

    return Positioned(
      right: 0.0,
      height: ScreenHeight(context) / 2,
      top: ScreenHeight(context) / 4,
      width: 120,
      child: Row(
        children: <Widget>[
          Container(
            alignment: Alignment(0, _indicatorY),
            width: 100,
            child: _indocatorHidden
                ? null
                : Stack(
                    alignment:
                        Alignment(-0.2, 0), //0, 0 是中心  顶部是 0，-1  左边中心是-1，0
                    children: <Widget>[
                      Image(
                        image: assetImg,
                        width: 60,
                      ),
                      Text(
                        _indicatorText,
                        style: TextStyle(fontSize: 25, color: Colors.white),
                      ),
                    ],
                  ), //气泡
          ),
          GestureDetector(
            child: Container(
              width: 20,
              color: _bkColor,
              child: Column(
                children: words,
              ),
            ),
            onVerticalDragUpdate: (DragUpdateDetails details) {
              int index = getIdex(context, details.globalPosition, _index_word);

              setState(() {
                _indicatorText = _index_word[index];
                //根据我们索引条的Alignment的Y值进行运算的。从 -1.1 到 1.1
                //整个的Y包含的值是2.2
                _indicatorY = 2.2 / _index_word.length * index - 1.1;
                _indocatorHidden = false;
              });
              widget.indexBarCallBack(_index_word[index]);
            }, //按住屏幕移动手指实时更新触摸的位置坐标

            onVerticalDragDown: (DragDownDetails details) {
              //globalPosition 自身坐标系
              int index = getIdex(context, details.globalPosition, _index_word);
              _indicatorText = _index_word[index];
              _indicatorY = 2.2 / _index_word.length * index - 1.1;
              _indocatorHidden = false;
              widget.indexBarCallBack(_index_word[index]);
              print('现在点击的位置是${details.globalPosition}');
              setState(() {
                _bkColor = Color.fromRGBO(1, 1, 1, 0.5);
                _textColor = Colors.white;
              });
            }, //触摸开始

            onVerticalDragEnd: (DragEndDetails details) {
              setState(() {
                _indocatorHidden = true;
                _bkColor = Color.fromRGBO(1, 1, 1, 0.0);
                _textColor = Colors.black;
              }); //触摸结束
            },
          ) //这个是索引条
        ],
      ),
    );
  }
}

// const INDEX_WORDS = [
//   '🔍',
//   '☆',
//   'A',
//   'B',
//   'C',
//   'D',
//   'E',
//   'F',
//   'G',
//   'H',
//   'I',
//   'J',
//   'K',
//   'L',
//   'M',
//   'N',
//   'O',
//   'P',
//   'Q',
//   'R',
//   'S',
//   'T',
//   'U',
//   'V',
//   'W',
//   'X',
//   'Y',
//   'Z'
// ];


lib/section_list.dart
import 'package:flutter/material.dart';
import 'custom_scroll_view_config.dart';

class SectionList extends StatefulWidget {
  ///每个section有几个item
  final List<int> itemCounts;
  final SectionHeaderBuilder sectionHeaderBuilder;
  final SectionListItemBuilder itemBuilder;
  final SectionListSeparatorBuilder separatorBuilder;
  final ValueChanged onTopChanged;
  final ValueChanged onEndChanged;
  final SectionListOffsetChanged onOffsetChanged;
  final double initialScrollOffset;
  final bool needSafeArea;

  SectionList(
      {@required this.itemCounts,
      @required this.sectionHeaderBuilder,
      @required this.itemBuilder,
      this.separatorBuilder,
      this.onTopChanged,
      this.onEndChanged,
      this.onOffsetChanged,
      this.initialScrollOffset = 0,
      this.needSafeArea = false,
      Key key})
      : assert(itemCounts != null, "itemCounts must not be null"),
        assert(sectionHeaderBuilder != null,
            "sectionHeaderBuilder must not be null"),
        assert(itemBuilder != null, "itemBuilder must not be null"),
        super(key: key);

  @override
  SectionListState createState() => SectionListState();
}

class SectionListState extends State<SectionList> {
  bool _top = true;
  bool _end = false;
  ScrollController _controller;

  ///index ： sectionIndex
  ///index和sectionIndex的对应关系
  Map<int, int> _headerIndexMap = {};

  jumpTo(double offset) {
    _controller.jumpTo(offset);
  }

  animateTo(
    double offset, {
    @required Duration duration,
    @required Curve curve,
  }) {
    _controller.animateTo(offset, duration: duration, curve: curve);
  }

  @override
  void initState() {
    super.initState();
    _controller =
        ScrollController(initialScrollOffset: widget.initialScrollOffset);
    if (widget.initialScrollOffset > 0) {
      _top = false;
    }
    _controller.addListener(() {
      double offset = _controller.offset;
      if (widget.onTopChanged != null) {
        bool top = offset <= _controller.position.minScrollExtent;
        if (_top != top) {
          _top = top;
          widget.onTopChanged(_top);
        }
      }
      if (widget.onEndChanged != null) {
        bool end = offset >= _controller.position.maxScrollExtent;
        if (_end != end) {
          _end = end;
          widget.onEndChanged(_end);
        }
      }
      if (widget.onOffsetChanged != null) {
        widget.onOffsetChanged(offset, _controller.position.maxScrollExtent);
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    _headerIndexMap.clear();
    return CustomScrollView(
      controller: _controller,
      physics: AlwaysScrollableScrollPhysics(),
      slivers: _buildSlivers(),
    );
  }

  List<Widget> _buildSlivers() {
    if (widget.needSafeArea) {
      return [SliverSafeArea(sliver: _buildSliverList())];
    } else {
      return [_buildSliverList()];
    }
  }

  _buildSliverList() {
    return SliverList(
        delegate: SliverChildBuilderDelegate((ctx, index) {
      if (_isHeaderIndex(index)) {
        int section = _headerIndexMap[index];
        return widget.sectionHeaderBuilder(ctx, section);
      } else {
        int headerIndex = _headerIndex(index);
        int section = _headerIndexMap[headerIndex];
        int sectionIndex = index - headerIndex;
        if (_isItemIndex(sectionIndex)) {
//              print(index);
          int fixIndex = (sectionIndex - 1) ~/ 2;
          return widget.itemBuilder(ctx, SectionIndexPath(section, fixIndex));
        } else {
          int fixIndex = (sectionIndex - 1) ~/ 2;
          bool isLast = fixIndex == widget.itemCounts[section] - 1;
          return widget.separatorBuilder(
              ctx, SectionIndexPath(section, fixIndex), isLast);
        }
      }
    }, childCount: _itemCounts()));
  }

  _itemCounts() {
    int ret = 0;
    int sectionIndex = 0;
    _headerIndexMap[0] = sectionIndex;
    widget.itemCounts.forEach((element) {
      //一个item，一个separator
      ret += element * 2;
      //一个header
      ret++;
      sectionIndex++;
      _headerIndexMap[ret] = sectionIndex;
    });
    return ret;
  }

  bool _isHeaderIndex(int index) {
    return _headerIndexMap[index] != null;
  }

  bool _isItemIndex(int sectionIndex) {
    return sectionIndex % 2 != 0;
  }

  //header 的index
  int _headerIndex(int index) {
    int ret = 0;
    while (index > 0) {
      if (_headerIndexMap[index] != null) {
        ret = index;
        break;
      }
      index--;
    }
    return ret;
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }
}


lib/hot_history_style.dart
import 'package:flutter/material.dart';

class HotHistoryStyle {
  final bool showHotHistory;
  final String hotText;
  final TextStyle hotTextStyle;
  // final Function clearHistoryFunc;
  final String historyText;
  final TextStyle historyTextStyle;
  final EdgeInsetsGeometry padding;
  final EdgeInsetsGeometry margin;
  final Color bgColor;
  final double borderRadius;
  final Widget clearBtn;
  // Widget clearBtn = Container(
  //   // width: 100,
  //   // height: 25,
  //   decoration:
  //       BoxDecoration(border: Border.all(color: Colors.black54, width: 1)),
  //   child: Row(
  //     mainAxisAlignment: MainAxisAlignment.center,
  //     children: <Widget>[Icon(Icons.delete), Text('清空历史记录')],
  //   ),
  // );
  HotHistoryStyle(
      {this.showHotHistory = true,
      this.hotText = '热搜',
      this.hotTextStyle =
          const TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
      // this.clearHistoryFunc,
      this.historyText = '搜索',
      this.historyTextStyle =
          const TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
      this.padding = const EdgeInsets.all(10),
      this.margin = const EdgeInsets.all(10),
      this.bgColor = const Color.fromRGBO(233, 233, 233, 0.9),
      this.borderRadius = 10,
      this.clearBtn = const Icon(Icons.delete)});
}


example/lib/main.dart
import 'dart:math';
import 'dart:ui';

import 'package:flappy_search_bar_fork/flappy_search_bar.dart';
import 'package:flappy_search_bar_fork/hot_history_style.dart';
import 'package:flappy_search_bar_fork/scaled_tile.dart';
import 'package:flappy_search_bar_fork/search_bar_style.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Home(),
    );
  }
}

// Size _getTextSize(String text, TextStyle style) {
//   final TextPainter textPainter = TextPainter(
//       text: TextSpan(text: text, style: style),
//       maxLines: 1,
//       textDirection: TextDirection.ltr)
//     ..layout(minWidth: 0, maxWidth: double.infinity);
//   return textPainter.size;
// }

class Post {
  final String title;
  final String body;

  Post(this.title, this.body);
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  final SearchBarController<Post, String> _searchBarController =
      SearchBarController();
  bool isReplay = false;
  Random random = Random();
  double cancellationWidgetWidth = 0;
  TextAlign textAligin;
  // FocusNode focusNode;
  // //<List<Post>>
  // List<String>>
  Future<List> _getALlPosts(String text, SearchType searchType) async {
    await Future.delayed(Duration(seconds: text.length == 4 ? 10 : 1));

    // List posts = [];

    // var random = new Random();
    // for (int i = 0; i < 10; i++) {
    //   posts.add("${text} : ${random.nextInt(100)}");
    // }
    // return List.from(posts);
    //

    List<Post> posts = [];

    var random = new Random();
    for (int i = 0; i < 10; i++) {
      posts.add(Post(" $i", "body random number : ${random.nextInt(100)}"));
    }
    return posts;
  }

//<List<Post>>
  Future<List<Post>> _getData(bool isMore) async {
    var text = 'fresh fresh';
    await Future.delayed(Duration(seconds: text.length == 4 ? 10 : 1));
    if (isReplay) return [Post("Replaying !", "Replaying body")];
    if (text.length == 5) throw Error();
    if (text.length == 6) return [];
    List<Post> posts = [];

    var random = new Random();
    for (int i = 0; i < 10; i++) {
      posts
          .add(Post("$text $i", "body random number : ${random.nextInt(100)}"));
    }
    return posts;
  }

  @override
  void initState() {
    super.initState();
    // focusNode = FocusNode();
    textAligin = TextAlign.center;
    cancellationWidgetWidth = 0;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Container(
            child: GestureDetector(
          behavior: HitTestBehavior.translucent,
          onTap: () {
            // 触摸收起键盘
            FocusScope.of(context).requestFocus(FocusNode());
            setState(() {
              // focusNode = FocusNode();
            });
          },
          child: SearchBar<Post, String>(
            searchBarPadding: EdgeInsets.symmetric(horizontal: 10),
            headerPadding: EdgeInsets.symmetric(horizontal: 10),
            listPadding: EdgeInsets.symmetric(horizontal: 10),
            getListData: _getData,
            searchType: SearchType.search,
            onChange: _getALlPosts,
            searchBarController: _searchBarController,
            textAligin: textAligin,
            textPaddingLeft: 30,
            // focusNode: focusNode,
            focusCall: () {
              setState(() {
                print('focuscall');
                textAligin = TextAlign.left;
                cancellationWidgetWidth = 120;
              });
            },
            hintText: '搜索',
            hintStyle: TextStyle(
              textBaseline: TextBaseline.alphabetic,
              fontSize: 16,
            ),
            placeHolder: Text("placeholder"),
            cancellationWidget: Text("Cancel"),
            cancellationWidgetWidth: cancellationWidgetWidth,
            emptyWidget: Text("empty"),
            hoverListProperty: HoverListProperty(itemCounts: [5]),
            scaledTileProperty: ScaledTitleProperty(
              indexedScaledTileBuilder: (int index) => ScaledTile.extent(1, 60),
              onItemFound: (post, int index) {
                return Container(
                  color: Colors.lightBlue,
                  child: ListTile(
                    title: Text((post.title)),
                    // isThreeLine: true,
                    subtitle: Text(post.body),
                    onTap: () {
                      Navigator.of(context).push(
                          MaterialPageRoute(builder: (context) => Detail()));
                    },
                  ),
                );
              },
            ),
            // indexedScaledTileBuilder: (int index) =>ScaledTile.extent(1, 60),
            //indexedScaledTileBuilder: (int index) => ScaledTile.count(1, index.isEven ? 2 : 1),

            header: Row(
              children: <Widget>[
                RaisedButton(
                  child: Text("Sort"),
                  onPressed: () {
                    _searchBarController.sortList((Post a, Post b) {
                      return a.body.compareTo(b.body);
                    });
                  },
                ),
                RaisedButton(
                  child: Text("Desort"),
                  onPressed: () {
                    _searchBarController.removeSort();
                  },
                ),
                RaisedButton(
                  child: Text("Replay"),
                  onPressed: () {
                    isReplay = !isReplay;
                    _searchBarController.replayLastSearch();
                  },
                ),
                RaisedButton(
                    child: Text("Random"),
                    onPressed: () {
                      // cancellationWidgetWidth = 0; //random.nextInt(250) + 50.0;
                      // textPaddingLeft = 250;
                      // focusNode = FocusNode();
                      setState(() {});
                    })
              ],
            ),
            onCancelled: (String value) {
              print("Cancelled triggered");

              setState(() {
                textAligin = TextAlign.center;
                cancellationWidgetWidth = 0; //random.nextInt(250) + 50.0;
              });
            },
            mainAxisSpacing: 0,
            crossAxisSpacing: 10,
            crossAxisCount: 1,
            onItemLenovoFound: (item, int index) {
              return Text('item ---- ${index}');
            }, //2,
          ),
        )),
      ),
    );
  }
}

class Detail extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Column(
          children: <Widget>[
            IconButton(
              icon: Icon(Icons.arrow_back),
              onPressed: () => Navigator.of(context).pop(),
            ),
            Text("Detail"),
          ],
        ),
      ),
    );
  }
}
